src/units/abilities.cpp

```diff
@@ -1168,33 +1168,92 @@ void attack_type::modified_attacks(unsigned & min_attacks,
 	}
 }
 
-//Functions used for change damage_type list with damage
-static std::optional<std::string> select_damage_type(const unit_ability_list& abil_list, const std::string& type)
+//functions for select damage type who must be returned for each attribute used in damage_type()
+static std::vector<std::string> select_type(const unit_ability_list& abil_list, const std::string& key_name)
```

Comment:
I don't really understand why you keep alternate and replacement type logic mixed.

There should be one function that picks the best replacement type, one function that picks the best alternate type and one function that simply returns a list of all alternate types. The last function should be called from the report and the report code should decide to limit the number of types.

Reply:
> I don't really understand why you keep alternate and replacement type logic mixed.
> 
> There should be one function that picks the best replacement type, one function that picks the best alternate type and one function that simply returns a list of all alternate types. The last function should be called from the report and the report code should decide to limit the number of types.

Code separated